feat: Initial work on rewriting closures to regular functions with hi…

alehander92 opened the pull request:
# Description

This commit implements the following mechanism:

On a line where a lambda expression is encountered, we initialize a tuple for the captured lambda environment and we rewrite the lambda to a regular function taking this environment as an additional parameter. All calls to the closure are then modified to insert this hidden parameter.

In other words, the following code:

```
let x = some_value;
let closure = |a| x + a;

println(closure(10));
println(closure(20));
```

is rewritten to:

```
fn closure(env: (Field,), a: Field) -> Field {
  env.0 + a
}

let x = some_value;
let closure_env = (x,);

println(closure(closure_env, 10));
println(closure(closure_env, 20));
```

In the presence of nested closures, we propagate the captured variables implicitly through all intermediate closures:

```
let x = some_value;

# The hidden env here will include `x` even through
# it's not directly used by the outer closure:
let closure = |a, c|
  # here, `x` is initialized from the hidden env of the outer closure
  let inner_closure = |b|
    a + b + x

  inner_closure(c)
```

To make these transforms possible, the following changes were made to the logic of the HIR resolver and the monomorphization pass:

* In the HIR resolver pass, the code determines the precise list of variables captured by each lambda. Along with the list, we compute the index of each captured var within the parent closure's environment (when the capture is propagated).

* Introduction of a new `Closure` type in order to be able to recognize the call-sites that need the automatic environment variable treatment. It's a bit unfortunate that the Closure type is defined within the `AST` modules that are used to describe the output of the monomorphization pass, because we aim to eliminate all closures during the pass. A better solution would have been possible if the type check pass after HIR resolution was outputting types specific to the HIR pass (then the closures would exist only within this separate non-simplified type system).

* The majority of the work is in the Lambda processing step in the monomorphizer which performs the necessary transformations based on the above information.

Remaining things to do:

* There are a number of pending TODO items for various minor unresolved loose ends in the code.

* There are a lot of possible additional tests to be written.

* Update docs


## Problem\*

Resolves https://github.com/noir-lang/noir/issues/1254

## Documentation

- [x] This PR requires documentation updates when merged.

  <!-- If checked, check one of the following: -->

  - [x] I will submit a noir-lang/docs PR.

  <!-- Submit a PR on https://github.com/noir-lang/docs. Thank you! -->

  - [ ] I will request for and support Dev Rel's help in documenting this PR.

  <!-- List / highlight what should be documented. -->
  <!-- Dev Rel will reach out for clarifications when needed. Thank you! -->

# PR Checklist\*

- [ ] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.
: `cargo fmt`

jfecher commented on crates/noirc_frontend/src/monomorphization/ast.rs:
```diff
@@ -223,6 +229,52 @@ impl Type {
+        Expression::For(_for_expr) => unreachable!("TODO: How do we get the type of a for loop?"),
+        Expression::If(if_expr) => if_expr.typ.clone(),
+        Expression::Tuple(elements) => Type::Tuple(elements.iter().map(type_of).collect()),
+        Expression::ExtractTupleField(tuple, index) => match tuple.as_ref() {
+            Expression::Tuple(fields) => type_of(&fields[*index]),
+            _ => unreachable!("ICE: Tuple field access on non-tuple type"),
+        },
```
This may be a non-tuple if it is e.g. a variable: `foo.field` or `foo.0`. If we need the type here we'd need to store it in this variant.

alehander92 replied:
type_of removed for now

jfecher commented on crates/noirc_frontend/src/monomorphization/ast.rs:
```diff
@@ -223,6 +229,52 @@ impl Type {
+        Expression::Let(let_stmt) => type_of(let_stmt.expression.as_ref()),
+        Expression::Constrain(contraint, _) => type_of(contraint.as_ref()),
+        Expression::Assign(assign) => type_of_lvalue(&assign.lvalue),
+        Expression::Semi(expr) => type_of(expr.as_ref()), // TODO: Is this correct?
```
A semicolon expression always has the type `Type::Unit`

alehander92 replied:
fixed

jfecher commented on crates/noirc_frontend/src/hir/type_check/expr.rs:
```diff
@@ -256,7 +264,9 @@ impl<'interner> TypeChecker<'interner> {
                 });
-                Type::Function(params, Box::new(lambda.return_type))
+
+                let function_type = Type::Function(params, Box::new(lambda.return_type));
+                Type::Closure(Box::new(function_type))
```
This approach with adding the environment parameters to the actual function type and wrapping that in a Closure type wrapper seems a bit odd/roundabout to me. Can you explain why you took this approach?

alehander92 replied:
now using just a `Function` variant

jfecher reviewed (commented):
I've yet to go through the whole PR so far, but here's an initial review of mostly the type system related parts. Tomorrow I'll revisit this and go through the name resolution and monomorphization changes.

alehander92 commented on crates/noirc_frontend/src/hir/resolution/resolver.rs:
```diff
@@ -478,24 +477,24 @@ impl<'a> Resolver<'a> {
     }
 
-    fn get_ident_from_path(&mut self, path: Path) -> HirIdent {
+    fn get_ident_from_path(&mut self, path: Path) -> (HirIdent, usize) {
```
find_variable can return non 0 i guess

alehander92 commented on crates/noirc_frontend/src/hir/resolution/resolver.rs:
```diff
@@ -889,7 +891,52 @@ impl<'a> Resolver<'a> {
+                        DefinitionKind::GenericType(_) => {}
+                        // We ignore the above definition kinds because only local variables can be captured by closures.
+                        DefinitionKind::Local(_) => {
+                            let mut transitive_capture_index: Option<usize> = None;
```
sounds reasonable, it is deeply nested indeed

alehander92 replied:
refactored out

alehander92 commented on crates/noirc_frontend/src/monomorphization/ast.rs:
```diff
@@ -223,6 +229,52 @@ impl Type {
+        Expression::Binary(_binary) => unreachable!("TODO: How do we get the type of a Binary op"),
+        Expression::Index(index) => index.element_type.clone(),
+        Expression::Cast(cast) => cast.r#type.clone(),
+        Expression::For(_for_expr) => unreachable!("TODO: How do we get the type of a for loop?"),
```
ok, makes sense

jfecher commented on crates/noirc_frontend/src/monomorphization/mod.rs:
```diff
@@ -953,27 +1000,79 @@ impl<'interner> Monomorphizer<'interner> {
+                }
+            }
+        }));
+        let env_typ = ast::type_of(&env_tuple);
```
If we stored the env type on the lambda from type checking we could get rid of this type_of call

alehander92 replied:
we do store it, now with passing `expr` it seems we can access it

jfecher commented on crates/noirc_frontend/src/monomorphization/mod.rs:
```diff
@@ -953,27 +1000,79 @@ impl<'interner> Monomorphizer<'interner> {
+        }));
+        let env_typ = ast::type_of(&env_tuple);
+
+        let env_let_stmt = ast::Expression::Let(ast::Let {
+            id: env_local_id,
+            mutable: true,
+            name: env_name.to_string(),
+            expression: Box::new(env_tuple),
+        });
```
Why is the env mutable?

jfecher commented on crates/noirc_frontend/src/monomorphization/mod.rs:
```diff
@@ -953,27 +1000,79 @@ impl<'interner> Monomorphizer<'interner> {
+
+        let mut parameters = vec![];
+        parameters.push((env_local_id, true, env_name.to_string(), env_typ));
+        parameters.extend(converted_parameters);
```
We don't use converted_parameters after this, so this can be an `append`

alehander92 commented on crates/noirc_frontend/src/hir_def/expr.rs:
```diff
@@ -191,9 +191,25 @@ impl HirBlockExpression {
+    /// in the same scope as the closure. In a closure-inside-another-closure
+    /// scenarios, we might have a transitive captures of variables that must
+    /// be propagated during the construction of each closure. In this case,
+    /// we store the index of the captured variable in the environment of our
+    /// direct parent closure. We do this in order to simplify the HIR to AST
+    /// transformation in the monomorphization pass.
+    pub transitive_capture_index: Option<usize>,
```
not exactly, it's index in the parent env, the lambdas propagate hidden variables, as in this example by Zahary:



```
let x = some_value;

# The hidden env here will include `x` even through
# it's not directly used by the outer closure:
let closure = |a, c|
  # here, `x` is initialized from the hidden env of the outer closure
  let inner_closure = |b|
    a + b + x

  inner_closure(c)

```

alehander92 commented on crates/noirc_frontend/src/hir/resolution/resolver.rs:
```diff
@@ -889,7 +891,52 @@ impl<'a> Resolver<'a> {
-                let hir_ident = self.get_ident_from_path(path);
+                let (hir_ident, var_scope_index) = self.get_ident_from_path(path);
+
+                if hir_ident.id != DefinitionId::dummy_id() {
```
i don't want to change it so much: however, i did refactor out a method as you suggested, thank you

kevaundray commented:
@alehander92 what is the status of this PR?

alehander92 commented:
@kevaundray we discussed some lang/impl questions that appeared while working on this PR with @jfecher through the last week, and I prepared some fixes for this PR and the initial closure conversion.(also rebased after upstream dropped the old ssa backend) (More advanced capabilities can be merged as a subsequent patch probably?): I'll try to push the current changes soon

alehander92 commented:
I've just pushed the rebased version of the original changes, soon I'll push some more fixes

alehander92 commented on crates/noirc_frontend/src/node_interner.rs:
```diff
@@ -634,5 +634,6 @@ fn get_type_method_key(typ: &Type) -> Option<TypeMethodKey> {
         | Type::Constant(_)
         | Type::Error
         | Type::Struct(_, _) => None,
+        Type::Closure(_) => None, // TODO: Is this correct? How do we add methods to functions? Can we do the same for closures?
```
now using just the `Function` variant

alehander92 commented on crates/noirc_frontend/src/monomorphization/ast.rs:
```diff
@@ -223,6 +229,52 @@ impl Type {
     }
 }
 
+pub fn type_of_lvalue(lvalue: &LValue) -> Type {
```
moved

alehander92 commented on crates/noirc_frontend/src/hir_def/types.rs:
```diff
@@ -74,6 +74,11 @@ pub enum Type {
+    /// A closure (a pair of a function pointer and a tuple of captured variables).
+    /// Stores the underlying function type, which has been modifies such that the
+    /// first parameter is the type of the captured variables tuple.
+    Closure(Box<Type>),
```
now using just the `Function` variant

alehander92 commented on crates/noirc_frontend/src/monomorphization/mod.rs:
```diff
@@ -739,17 +769,34 @@ impl<'interner> Monomorphizer<'interner> {
 
+    fn is_function_closure(&self, func: &ast::Expression) -> bool {
+        matches!(ast::type_of(func), ast::Type::Tuple(_))
+    }
```
ok, that's how we'll do it

alehander92 commented on crates/noirc_frontend/src/monomorphization/ast.rs:
```diff
@@ -223,6 +229,52 @@ impl Type {
+        },
+        Expression::Block(stmts) => type_of(stmts.last().unwrap()),
+        Expression::Unary(unary) => unary.result_type.clone(),
+        Expression::Binary(_binary) => unreachable!("TODO: How do we get the type of a Binary op"),
```
type_of removed for now

alehander92 commented:
Did a transformation to using ` Function` variant for closures too, included some tests, fixed some problems and tried to address most comments from the review. I found a regression in one case of returning a closure, however my understanding is advanced returning closure/correct array of closure capabilities can be added in a next MR

jfecher commented:
>  I found a regression in one case of returning a closure, however my understanding is advanced returning closure/correct array of closure capabilities can be added in a next MR

Does this mean we can't return closures at all? I also noticed the current tests mostly test free functions rather than closures with an environment. What cases are expected to be implemented by this PR?

jfecher commented:
Looks like this PR crashes for the input:
```rs
fn main() {
    let x: u32 = 32;

    let f = if x > 2 {
        || x
    } else {
        || x + 2342
    };

    dep::std::println(f());
}
```
Also, when trying:
```rs
use dep::std::println;

fn main() {
    let x: u32 = 32;

    let f = if x > 2 {
        || x
    } else {
        three
    };

    println(f());
}

fn three() -> u32 {
    3
}
```
I get the rather confusing error message:
```
error: Expected type fn((u32)) -> u32 [(u32)], found type fn() -> u32 [()]
   ┌─ /Users/jakefecher/Code/Noir/noir/short/src/main.nr:6:13
   │  
 6 │       let f = if x > 2 {
   │ ╭─────────────'
 7 │ │         || x
 8 │ │     } else {
 9 │ │         three
10 │ │     };
   │ ╰─────'
   │  
   = Expected the types of both if branches to be equal
```
Some suggestions to improve this:
1. It is confusing for users that the closure type includes its environment as a parameter. I think these should be kept separate in the type system. We should't need to prepend the environment parameter until monomorphization.
2. The enviornment type of the free function is displayed as `[()]`, we should probably just omit this entirely.
3. It's probably fine to display the closure environment as `[(u32)]` since there isn't much prior art here. Users who aren't familiar with the internals of closures here may be confused though. Maybe we'd benefit from being more explicit/verbose about it? E.g. `Expected type fn() -> u32 with environment (u32), found type fn() -> u32`

jfecher commented:
> Does this mean we can't return closures at all? I also noticed the current tests mostly test free functions rather than closures with an environment. What cases are expected to be implemented by this PR?

I tested just now and it seems we're able to return lambdas but not closures that capture an environment. I'd like if that functionality were part of this PR as well. At the same time, I don't want to step on your toes and make things more difficult since I'm aware keeping up with merge conflicts gets increasingly difficult the larger the PR. So let me know if this would be too much. I'll try to be more consistent in reviewing these PRs in a more timely manner as well.

alehander92 commented:
> ..
> Some suggestions to improve this:
> 
>     1. It is confusing for users that the closure type includes its environment as a parameter. I think these should be kept separate in the type system. We should't need to prepend the environment parameter until monomorphization.
> 
>     2. The enviornment type of the free function is displayed as `[()]`, we should probably just omit this entirely.
> 
>     3. It's probably fine to display the closure environment as `[(u32)]` since there isn't much prior art here. Users who aren't familiar with the internals of closures here may be confused though. Maybe we'd benefit from being more explicit/verbose about it? E.g. `Expected type fn() -> u32 with closure environment (u32), found type fn() -> u32`


I agree with 2 and 3: i've fixed those in the last commit

for 1: i've fixed the unify/display logic, so it should behave correctly in type checking, however part of the logic itself remains before the monomorphization pass: should I move this as well?

now, with those fixes, both examples fail with the same kind of ICE, I'll try to look into it.

Overally, it would be good to fix the returning of closures, but in my understanding it's deeply related to the implicit genericness of functions with closures and instantiating them, which we planned for the next MR, so i think it would be easier to separate them like that(if we remain with the 2 MR plan)

alehander92 commented:
I also updated the temporarily commented out `higher_order_functions` test assertion code which depended on the #2057 fix: I am not totally sure what behavior we want when we capture mutable variables, as it seems in Rust this usually leads to a borrowing error if we mutate them meanwhile.
With the MR it seems we capture the capture-time value and later mutations don't affect us, but I am not sure what the desired behavior is

jfecher commented:
@alehander92 the expected behavior of capturing mutable variables in a lambda should be copying them - so mutating the captured variable should not affect the original. If the variable is declared as a mutable reference `let x = &mut ...;` then the copy should still mutate the original. The good news is I don't think you'll need to do any extra work to preserve this behavior.

jfecher commented on crates/nargo_cli/tests/test_data/higher_order_fn_selector/src/main.nr:
```diff
@@ -0,0 +1,49 @@
+    1
+}
+
+fn selector(flag:&mut bool) -> fn(&mut Field) -> Field { //TODO: Can we have fn(&mut Field) -> () return type?
```
You should be able to, is there a bug?

alehander92 replied:
huh, I might have been confused, i improved the test now and used that

jfecher commented on crates/nargo_cli/tests/test_data/higher_order_functions/src/main.nr:
```diff
@@ -0,0 +1,84 @@
+
+    // When you copy mutable variables,
+    // the capture of the copies shouldn't change:
+    let mut x = 2;
+    x = x + 1;
+    let z = x;
```
I think a closure was removed here. Since #2057 is fixed and merged, can we restore the old test?

jfecher commented on crates/nargo_cli/tests/test_data/higher_order_functions/src/main.nr:
```diff
@@ -0,0 +1,84 @@
+    assert(evens.fold(0, |a, b| a + b) == 12);
+    assert(evens.reduce(|a, b| a + b) == 12);
+
+    // TODO: is this a sort_via issue with the new backend,
+    // or something more general?
+    //
+    // currently it fails only with `--experimental-ssa` with
+    //  "not yet implemented: Cast into signed"
+    // but it worked with the original ssa backend
+    // (before dropping it)
```
Can you open issues for any bugs found? This seems like an important one to fix if it's preventing `sort_via` from working.

alehander92 replied:
ok, opened https://github.com/noir-lang/noir/issues/2121

jfecher commented on crates/noirc_frontend/src/hir/type_check/expr.rs:
```diff
@@ -865,6 +872,38 @@ impl<'interner> TypeChecker<'interner> {
+        fn_ret: &Type,
+        callsite_args: &Vec<(Type, ExprId, Span)>,
+        span: Span,
+        skip_params: usize,
+    ) -> Type {
+        let real_fn_params_count = fn_params.len() - skip_params;
```
We should remove the `skip_params` as well, since we don't need the environment parameter in the function's parameter types. It can be added later during monomorphization when we need it there but there's not much use storing it in two places in the function type.

alehander92 replied:
fixed

jfecher commented on crates/noirc_frontend/src/hir/type_check/expr.rs:
```diff
@@ -881,38 +920,29 @@ impl<'interner> TypeChecker<'interner> {
+                        Type::Tuple(_) => {
+                            1 // closure env
+                        }
+                        _ => unreachable!(
+                            "function env internal type should be either Unit or Tuple"
+                        ),
```
It can be a type variable or named generic as well - but we can remove this check anyway when skip_params is removed

alehander92 replied:
you're right, removed

jfecher commented on crates/noirc_frontend/src/monomorphization/ast.rs:
```diff
@@ -102,6 +101,13 @@ pub struct Binary {
+pub struct Lambda {
+    pub function: Ident,
+    pub env: Ident,
+    pub typ: Type, // TODO: Perhaps this is not necessary
```
Was it necessary?

alehander92 replied:
it seems not: removing it

jfecher commented on crates/noirc_frontend/src/monomorphization/ast.rs:
```diff
@@ -319,9 +325,9 @@ impl std::fmt::Display for Type {
+            Type::Function(args, ret, env) => {
                 let args = vecmap(args, ToString::to_string);
-                write!(f, "fn({}) -> {}", args.join(", "), ret)
+                write!(f, "fn({}) -> {} [{}]", args.join(", "), ret, env)
```
For consistency we should also probably have the same "with closure environment" printout here as well

alehander92 replied:
ok, fixed(in next commit)

jfecher commented on crates/noirc_frontend/src/monomorphization/mod.rs:
```diff
@@ -686,12 +741,39 @@ impl<'interner> Monomorphizer<'interner> {
-        }))
+        let mut block_expressions = vec![];
+
+        let is_closure = self.is_function_closure(call.func);
+        if is_closure {
+            let extracted_func: ast::Expression;
+            let hir_call_func = self.interner.expression(&call.func);
+            if let HirExpression::Lambda(l) = hir_call_func {
+                let (setup, closure_variable) = self.lambda_with_setup(l, call.func);
+                block_expressions.push(setup);
+                extracted_func = closure_variable;
+            } else {
+                extracted_func = *original_func;
+            }
+            func = Box::new(ast::Expression::ExtractTupleField(
+                Box::new(extracted_func.clone()),
+                1usize,
+            ));
+            let env_argument = ast::Expression::ExtractTupleField(Box::new(extracted_func), 0usize);
+            arguments.insert(0, env_argument);
+        } else {
+            func = original_func.clone();
+        };
```
Can we break this out into another function?

alehander92 replied:
I am not fully sure if this logic wouldn't be more seriously improved anyway in the next reform, and currently it updates several local variables(`block_expressions`, `arguments`), so I'd prefer to leave the refactoring out for the next PR: otherwise I agree

jfecher commented on crates/noirc_frontend/src/monomorphization/mod.rs:
```diff
@@ -877,27 +977,109 @@ impl<'interner> Monomorphizer<'interner> {
+        });
+
+        (block_let_stmt, closure_ident)
+        // ast::Expression::Block(vec![block_let_stmt, closure_ident])
```
```suggestion
```

jfecher reviewed (commented):
Some more suggestions before we can merge. I think the most important one is removing `skip_params` since the code will be cleaner without it. We can leave any larger items for later PRs.

alehander92 commented:
@jfecher thanks, I'll try to address them.

About the mutable references: yes, mutations of the variable that they point to from the outside affect the captured variable behavior, however we currently get a lookup ICE if we try to set the variable itself from the closure, e.g.

```noir
    let closure_updating_captured_mutable_reference = (|y| {
        *ref_x = y;
    });
```

(also, if we try to directly set `x = y`;)

I assume the first should work and the second: not entirely sure, probably not, as we are capturing a copy, so: a friendly typecheck error?

jfecher commented:
@alehander92 hmm, I would expect it to work as well to mutate the variable itself from inside the closure. Assuming the code is something like:
```rs
let x = &mut 42;
let set_x = |y| { x = y; };

assert(*x == 42);
set_x(44);
assert(*x == 44);
set_x(*x);
assert(*x == 44);
```
I would expect it to work without type errors as well - since Noir allows multiple aliasing mutable references. Can you open an issue for this panic and we can address it in a later PR? It sounds like it might have more to do with mutable references being bugged than closures.

alehander92 commented:
@jfecher I assume you mean `*x = y`: yes,it seems it also crashes on master indeed, that's surprising to me, I usually assume I introduce all the issues

I'll open an issue for it

alehander92 commented on crates/noirc_frontend/src/hir_def/types.rs:
```diff
@@ -624,7 +627,7 @@ impl Type {
                 fields.iter().any(|field| field.contains_numeric_typevar(target_id))
             }
-            Type::Function(parameters, return_type) => {
+            Type::Function(parameters, return_type, _env) => {
```
ok, checking env too

alehander92 commented on crates/noirc_frontend/src/hir/type_check/expr.rs:
```diff
@@ -881,38 +920,29 @@ impl<'interner> TypeChecker<'interner> {
                 let ret = self.interner.next_type_variable();
                 let args = vecmap(args, |(arg, _, _)| arg);
-                let expected = Type::Function(args, Box::new(ret.clone()));
+                let expected = Type::Function(args, Box::new(ret.clone()), Box::new(Type::Unit));
```
Ok, did

```diff
-                let expected = Type::Function(args, Box::new(ret.clone()), Box::new(Type::Unit));
+                let env_type = self.interner.next_type_variable();
+                let expected = Type::Function(args, Box::new(ret.clone()), Box::new(env_type));
```

I hope this is correct

alehander92 commented on crates/noirc_frontend/src/monomorphization/mod.rs:
```diff
@@ -867,7 +949,25 @@ impl<'interner> Monomorphizer<'interner> {
     }
 
-    fn lambda(&mut self, lambda: HirLambda) -> ast::Expression {
+    fn lambda_with_setup(
```
I'd prefer if this could also be refactored in next PR(as you said, that other bigger subparts can be left for the next one)

alehander92 commented:
@jfecher : i removed the skip_params/additional params handling from typechecking, tried to address most of the issues, except the refactoring ones, also Alex from our team fixed lambdas-without-captured-variables handling, as they don't need to really go through the closure transformation(this additional case can be a part of the eventual lambda/lambda_with_setup refactoring too)

I also found the root of the ICE in one of your examples (`let f = if .. || x .. else .. || x + ..; f() `)

it's because we set the `If` type during the monomorphization using `Self::convert_type(&self.interner.id_type(expr));`: so we get a `fn` type, even if it's actually a rewritten to a `tuple` closure. I tested manually that assigning the tuple type seems to fix at least this example, so something like `type_of` would help here.

Is there a reason why we can't have `typ` field for each expression, which IMHO seems like a simpler/more predictable solution than a `type_of`? I can imagine that one might be wary of bugs: discrepancies between assigned types and the actual correct types(and to expect that with `type_of` one would get "correct by construction" types), but this doesn't seem as such a critical concern. Is it because of having different data structures for some passes, or performance?

jfecher commented:
> I am not sure if the ICE you found is a blocker for this PR, so here is the research for it:

Since it's closure related, we can leave it to a later PR.

> it's because we set the If type during the monomorphization using `Self::convert_type(&self.interner.id_type(expr));`: so we get a fn type, even if it's actually a rewritten to a tuple closure. I tested manually that assigning the tuple type seems to fix at least this example, so something like type_of would help here.

I would think if we're converting closures into tuples, then `Self::convert_type` should return a tuple then for closures instead of a function type. Perhaps this can be fixed in a later PR as well.

> Is there a reason why we can't have typ field for each expression, which IMHO seems like a simpler/more predictable solution than a type_of? I can imagine that one might be wary of bugs: discrepancies between assigned types and the actual correct types(and to expect that with type_of one would get "correct by construction" types), but this doesn't seem as such a critical concern. Is it because of having different data structures for some passes, or performance?

I'm generally against adding a `type_of` method since it repeats the work of the type checker and although there is not a `typ` field directly on each expression, there is a type for every expression already. When type checking we store every expression's type in the `NodeInterner`, keyed by its expression ID. You can use `interner.id_type(id)` to access it. You should be able to find examples of this elsewhere in the monomorphizaiton pass as well.

jfecher commented on crates/noirc_frontend/src/hir_def/types.rs:
```diff
@@ -1192,9 +1202,9 @@ impl Type {
             }
 
-            (Function(params_a, ret_a), Function(params_b, ret_b)) => {
+            (Function(params_a, ret_a, _env_a), Function(params_b, ret_b, _env_b)) => {
```
We should still unify and subtype function environments, lets add this to the list of things for a later PR.

jfecher commented on crates/noirc_frontend/src/hir/type_check/expr.rs:
```diff
@@ -279,7 +279,13 @@ impl<'interner> TypeChecker<'interner> {
+                let env_type: Type =
+                    if captured_vars.is_empty() { Type::Unit } else { Type::Tuple(captured_vars) };
+
+                let params: Vec<Type> = vecmap(lambda.parameters, |(pattern, typ)| {
```
Ah, thought I'd be able to quickly commit this change before merging, guess not.

alehander92 replied:
fixed locally, comitted

jfecher reviewed (approved):
Thanks for the PR! The major items have been fixed and I've been holding this up long enough, so as we've discussed, let's leave the rest for a future PR.

jfecher commented:
An incomplete list of some of the TODO items, most are just style concerns:
- Unneeded type annotation https://github.com/noir-lang/noir/pull/1959#discussion_r1282240622
- Unify & subtype env https://github.com/noir-lang/noir/pull/1959#discussion_r1282243011
- Simplifying / breaking up `lambda_with_setup`
- Simplifying https://github.com/noir-lang/noir/pull/1959#discussion_r1281013486

Functionality:
- Returning closures
- Fixing any panics
- A scheme was discussed to implicitly add generics to functions/structs to get closures with different environments to work together. We may want to save this for the last step and have closures with incompatible generics (like rust's `impl Fn()`) first since it seems like a large / difficult change.

alehander92 commented:
@jfecher sorry, had to fix one test again: now they all pass locally (at least `cargo test --release --worskpace`), I hope they pass the CI as well this time.

I also removed the `: Vec<Type>` `params` annotation

alehander92 commented:
> I would think if we're converting closures into tuples, then Self::convert_type should return a tuple then for closures instead of a function type. Perhaps this can be fixed in a later PR as well. 

hm , we can try that in the new work

Savio-Sou commented:
Did this go into v0.10.0? It doesn't seem to show up in the changelogs 🤔

alexvitkov commented:
There's few more things related to this PR that we'd like to get merged first:
1. This syntax change: https://github.com/noir-lang/noir/pull/2383
2. Change the signatures of some standard library functions, so they work with closures https://github.com/noir-lang/noir/issues/2385#issuecomment-1686668621

Those are both small changes so it should be done today/tomorrow, after that I'll send a PR to docs describing the changes related to closures & give some examples on how to use them.

alexvitkov commented:
The changes in Noir have been merged, I've submitted a PR for docs: https://github.com/noir-lang/docs/pull/356
